model/jpa/src/main/java/org/keycloak/models/jpa/UserAdapter.java

```diff
@@ -392,7 +392,7 @@ public long getGroupsCountByNameContaining(String search) {
 
     @Override
     public void joinGroup(GroupModel group) {
-        if (isMemberOf(group)) return;
+        if (getGroupsStream().anyMatch(g -> group.getId().equals(g.getId()))) return;
```

Comment:
This change should also be propagated to https://github.com/keycloak/keycloak/blob/fd5a423b551b9836bda44d47b38d7cd34f8d27c4/model/map/src/main/java/org/keycloak/models/map/user/MapUserAdapter.java#L266-L269

Reply:
Actually, this one is a bit tricky. In `UserAdapter.java` this method was not adapted, it is still checking for transitive group membership. It is only not used anymore when joining a group. However, in `MapUserAdapter.java`I can see there is actually no check at all when joining a group: https://github.com/keycloak/keycloak/blob/52488150917c5c7dd68e7eef7eab7c66275f0e49/model/map/src/main/java/org/keycloak/models/map/user/MapUserAdapter.java#L257

I am not sure this is really intended.

Reply:
My bad, apologies. I marked wrong lines in the `MapUserAdapter` and you have correctly spotted that.
I meant propagating it to the `MapUserAdapter.joinGroup()` method that you have referenced.

Reply:
I added the check to `MapUserAdapter.java`.